codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt

```diff
@@ -1,56 +1,304 @@
-                *codegenScope, "target" to symbolProvider.toSymbol(inputShape)
+                "pub fn $fnName(${writerMut}writer: $writerType, input: &#{Input})",
+                "Input" to structureSymbol,
+                *codegenScope
             ) {
-                // TODO: Implement query operation serializer
-                writer.rust("unimplemented!()")
+                context.copy(writerExpression = "writer", valueExpression = ValueExpression.Reference("input"))
+                    .also { inner ->
+                        val members = inner.shape.members()
+                        if (members.isEmpty()) {
+                            rust("let (_, _) = (writer, input);") // Suppress unused argument warnings
+                        }
+                        for (member in members) {
+                            val memberContext = MemberContext.structMember(inner, member, symbolProvider)
+                            structWriter(memberContext) { writerExpression ->
+                                serializeMember(memberContext.copy(writerExpression = writerExpression))
+                            }
+                        }
+                    }
             }
         }
+        val borrowType = when (topLevel) {
+            true -> "&mut "
+            else -> ""
+        }
+        rust("#T($borrowType${context.writerExpression}, ${context.valueExpression.name});", structureSerializer)
     }
 
-    override fun documentSerializer(): RuntimeType {
-        TODO("AwsQuery doesn't support document types")
+    private fun RustWriter.serializeMember(context: MemberContext) {
+        val targetShape = model.expectShape(context.shape.target)
+        if (symbolProvider.toSymbol(context.shape).isOptional()) {
+            safeName().also { local ->
+                rustBlock("if let Some($local) = ${context.valueExpression.asRef()}") {
+                    val innerContext = context.copy(valueExpression = ValueExpression.Reference(local))
+                    serializeMemberValue(innerContext, targetShape)
+                }
+            }
+        } else {
+            serializeMemberValue(context, targetShape)
+        }
+    }
+
+    private fun RustWriter.serializeMemberValue(context: MemberContext, target: Shape) {
+        val writer = context.writerExpression
+        val value = context.valueExpression
+        when (target) {
+            is StringShape -> when (target.hasTrait<EnumTrait>()) {
+                true -> rust("$writer.string(${value.name}.as_str());")
+                false -> rust("$writer.string(${value.name});")
+            }
+            is BooleanShape -> rust("$writer.boolean(${value.asValue()});")
+            is NumberShape -> {
+                val numberType = when (symbolProvider.toSymbol(target).rustType()) {
+                    is RustType.Float -> "Float"
+                    is RustType.Integer -> "NegInt"
```

Comment:
is this right?

Reply:
ohhh it generates i64? Can you comment that?

Reply:
Added comment.